File: backend/app/schemas/vehicle.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Generic, TypeVar

from pydantic import BaseModel, field_validator

from app.models.vehicle import FuelType, Transmission, VehicleCondition, VehicleStatus
from app.schemas.vehicle_image import VehicleImageRead

T = TypeVar("T")


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int
    pages: int


class VehicleCreate(BaseModel):
    brand: str
    model: str
    year: int
    version: str | None = None
    color: str
    mileage: int
    fuel_type: FuelType
    transmission: Transmission
    condition: VehicleCondition
    body_type: str | None = None
    plate: str | None = None
    price_resale: Decimal
    price_public: Decimal
    description: str | None = None
    status: VehicleStatus = VehicleStatus.AVAILABLE

    @field_validator("year")
    @classmethod
    def valid_year(cls, v: int) -> int:
        if v < 1900 or v > 2100:
            raise ValueError("Year out of range")
        return v

    @field_validator("mileage")
    @classmethod
    def non_negative_mileage(cls, v: int) -> int:
        if v < 0:
            raise ValueError("Mileage cannot be negative")
        return v


class VehicleUpdate(BaseModel):
    brand: str | None = None
    model: str | None = None
    year: int | None = None
    version: str | None = None
    color: str | None = None
    mileage: int | None = None
    fuel_type: FuelType | None = None
    transmission: Transmission | None = None
    condition: VehicleCondition | None = None
    body_type: str | None = None
    plate: str | None = None
    price_resale: Decimal | None = None
    price_public: Decimal | None = None
    description: str | None = None
    status: VehicleStatus | None = None


class VehicleStatusUpdate(BaseModel):
    status: VehicleStatus


class VehicleFilters(BaseModel):
    brand: str | None = None
    model: str | None = None
    year_min: int | None = None
    year_max: int | None = None
    fuel_type: FuelType | None = None
    transmission: Transmission | None = None
    condition: VehicleCondition | None = None
    status: VehicleStatus | None = VehicleStatus.AVAILABLE
    company_id: uuid.UUID | None = None
    page: int = 1
    page_size: int = 20


class VehicleRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    company_id: uuid.UUID
    company_name: str
    brand: str
    model: str
    year: int
    version: str | None
    color: str
    mileage: int
    fuel_type: FuelType
    transmission: Transmission
    condition: VehicleCondition
    body_type: str | None
    plate: str | None = None
    price_resale: Decimal
    price_public: Decimal
    description: str | None
    status: VehicleStatus
    share_token: uuid.UUID
    images: list[VehicleImageRead]
    is_favorite_company: bool = False
    created_at: datetime


class VehicleListItem(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    company_id: uuid.UUID
    company_name: str
    brand: str
    model: str
    year: int
    version: str | None
    color: str
    mileage: int
    fuel_type: FuelType
    transmission: Transmission
    condition: VehicleCondition
    price_resale: Decimal
    price_public: Decimal
    status: VehicleStatus
    primary_image_url: str | None = None
    is_favorite_company: bool = False
    created_at: datetime


class VehiclePublic(BaseModel):
    """Returned for shareable links — no prices, no company info."""

    model_config = {"from_attributes": True}

    brand: str
    model: str
    year: int
    version: str | None
    color: str
    mileage: int
    fuel_type: FuelType
    transmission: Transmission
    condition: VehicleCondition
    body_type: str | None
    description: str | None
    images: list[VehicleImageRead]

File: frontend/src/types/vehicle.ts
export type FuelType = "gasoline" | "diesel" | "electric" | "hybrid" | "gnc";
export type Transmission = "manual" | "automatic";
export type VehicleCondition = "new" | "used";
export type VehicleStatus = "available" | "reserved" | "sold" | "pre_toma";

export interface VehicleImage {
  id: string;
  s3_key: string;
  url: string;
  display_order: number;
  is_primary: boolean;
}

export interface Vehicle {
  id: string;
  company_id: string;
  company_name: string;
  brand: string;
  model: string;
  year: number;
  version: string | null;
  color: string;
  mileage: number;
  fuel_type: FuelType;
  transmission: Transmission;
  condition: VehicleCondition;
  body_type: string | null;
  plate: string | null;
  price_resale: number;
  price_public: number;
  description: string | null;
  status: VehicleStatus;
  share_token: string;
  images: VehicleImage[];
  is_favorite_company: boolean;
  created_at: string;
  updated_at: string;
}

export interface VehicleListItem {
  id: string;
  company_id: string;
  company_name: string;
  brand: string;
  model: string;
  year: number;
  version: string | null;
  color: string;
  mileage: number;
  fuel_type: FuelType;
  transmission: Transmission;
  condition: VehicleCondition;
  price_resale: number;
  price_public: number;
  status: VehicleStatus;
  is_favorite_company: boolean;
  primary_image_url: string | null;
}

export interface VehiclePublic {
  brand: string;
  model: string;
  year: number;
  version: string | null;
  color: string;
  mileage: number;
  fuel_type: FuelType;
  transmission: Transmission;
  condition: VehicleCondition;
  body_type: string | null;
  description: string | null;
  images: VehicleImage[];
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface VehicleFilters {
  brand?: string;
  model?: string;
  year_min?: number;
  year_max?: number;
  fuel_type?: FuelType;
  transmission?: Transmission;
  condition?: VehicleCondition;
  status?: VehicleStatus;
  company_id?: string;
  page?: number;
  page_size?: number;
}

export interface VehicleCreate {
  brand: string;
  model: string;
  year: number;
  version?: string;
  color: string;
  mileage: number;
  fuel_type: FuelType;
  transmission: Transmission;
  condition: VehicleCondition;
  body_type?: string;
  plate?: string;
  price_resale: number;
  price_public: number;
  description?: string;
  status?: VehicleStatus;
}

export type VehicleUpdate = Partial<VehicleCreate>;

export interface UploadUrlResponse {
  upload_url: string;
  s3_key: string;
}
